File: rust/sdk/synor/src/resources/live_map.rs
//! In-memory live key-value map.
//!
//! [`LiveMap`] bridges live data-*producing* logic and live data-*consuming*
//! logic within a single Synor session. The producing side declares
//! `(key, value)` entries as **target states** via [`LiveMap::declare_entry`];
//! the consuming side reads it as a [`LiveMapView`] through
//! [`Ctx::mount_each_live`](crate::Ctx::mount_each_live). All data is held in an
//! in-process map that the engine keeps in sync through normal target-state
//! ownership, and that same map is exposed as a live source for downstream
//! components.
//!
//! Designed for live mode (`UpdateOptions::live`). An entry exists as long as
//! some live component declares it; when its declaring component stops declaring
//! it (or disappears), the entry is removed.
//!
//! ```ignore
//! let lm: LiveMap<String> = LiveMap::create(&ctx).await?; // inside the component tree
//! lm.declare_entry(&ctx, key, value)?;                    // producer: inside any component
//! ctx.mount_each_live(lm.clone(), process_entry).await?;  // consumer: one component per entry
//! ```

use std::collections::HashMap;
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::{Arc, Mutex};

use async_trait::async_trait;
use serde::de::DeserializeOwned;
use serde::{Deserialize, Serialize};
use tokio::sync::mpsc;

use crate::ctx::Ctx;
use crate::error::{Error, Result};
use crate::live_component::{LiveMapFeed, LiveMapSubscriber, LiveMapView};
use crate::target_state::{
    ChildTargetDef, StableKey, TargetAction, TargetActionSink, TargetChildInvalidation,
    TargetHandler, TargetReconcileOutput, TargetStateProvider, declare_target_state, mount_target,
    register_root_target_states_provider,
};

/// Process-wide counter giving each [`LiveMap`] a distinct container provider
/// name. Like Python's per-instance UUID, the name is session-scoped — LiveMap
/// is an in-memory, live-mode primitive, not cross-run persistent state.
static NEXT_ID: AtomicU64 = AtomicU64::new(0);

/// The bound a [`LiveMap`] value must satisfy: it is a target-state value
/// (serializable), is cloned into the in-memory map and the consumer, and is
/// compared by `==` to suppress no-op change notifications. Auto-implemented for
/// every qualifying type — you never implement it by hand.
pub trait LiveMapValue:
    Serialize + DeserializeOwned + Clone + PartialEq + Send + Sync + 'static
{
}
impl<V: Serialize + DeserializeOwned + Clone + PartialEq + Send + Sync + 'static> LiveMapValue
    for V
{
}

enum Change<V> {
    Upsert(String, V),
    Delete(String),
}

struct LiveMapInner<V> {
    entries: Mutex<HashMap<String, V>>,
    /// Always live, so producer changes queue even before a consumer attaches.
    sender: mpsc::UnboundedSender<Change<V>>,
    /// Taken by the single active `watch()`.
    receiver: Mutex<Option<mpsc::UnboundedReceiver<Change<V>>>>,
}

impl<V: LiveMapValue> LiveMapInner<V> {
    /// Apply one entry action to the in-memory map, emitting a change to the
    /// watcher only when the value actually changed (the `==` gate).
    fn apply(&self, key: String, value: Option<V>) {
        let mut entries = self.entries.lock().unwrap();
        match value {
            Some(v) => {
                if entries.get(&key) != Some(&v) {
                    entries.insert(key.clone(), v.clone());
                    let _ = self.sender.send(Change::Upsert(key, v));
                }
            }
            None => {
                if entries.remove(&key).is_some() {
                    let _ = self.sender.send(Change::Delete(key));
                }
            }
        }
    }
}

/// An in-memory, keyed collection that is both a target (declare entries into it)
/// and a [`LiveMapView`] (consume it with [`Ctx::mount_each_live`]). See the
/// [module docs](self). Cheap to clone — the producer and consumer share state.
pub struct LiveMap<V> {
    inner: Arc<LiveMapInner<V>>,
    entry_provider: TargetStateProvider<V>,
}

impl<V> Clone for LiveMap<V> {
    fn clone(&self) -> Self {
        Self {
            inner: self.inner.clone(),
            entry_provider: self.entry_provider.clone(),
        }
    }
}

impl<V: LiveMapValue> LiveMap<V> {
    /// Create a LiveMap and mount its backing target. Call inside a component
    /// context (an `App::update`/`run` closure or a mounted component).
    pub async fn create(ctx: &Ctx) -> Result<Self> {
        let (sender, receiver) = mpsc::unbounded_channel();
        let inner = Arc::new(LiveMapInner {
            entries: Mutex::new(HashMap::new()),
            sender,
            receiver: Mutex::new(Some(receiver)),
        });
        let id = NEXT_ID.fetch_add(1, Ordering::Relaxed);
        let container_provider = register_root_target_states_provider(
            ctx,
            format!("synor/livemap/{id}"),
            ContainerHandler {
                inner: inner.clone(),
            },
        )?;
        let container = container_provider.target_state("container", ContainerSpec);
        let entry_provider = mount_target::<ContainerSpec, V>(ctx, container).await?;
        Ok(Self {
            inner,
            entry_provider,
        })
    }

    /// Declare an entry, owned by the calling component. Call inside a component
    /// context. The entry lives until its declaring component stops declaring it.
    pub fn declare_entry(&self, ctx: &Ctx, key: impl Into<String>, value: V) -> Result<()> {
        let key = key.into();
        declare_target_state(
            ctx,
            self.entry_provider
                .target_state(StableKey::Str(Arc::from(key.as_str())), value),
        )
    }
}

#[async_trait]
impl<V: LiveMapValue> LiveMapView<String, V> for LiveMap<V> {
    async fn scan(&self) -> Result<Vec<(String, V)>> {
        let entries = self.inner.entries.lock().unwrap();
        Ok(entries
            .iter()
            .map(|(k, v)| (k.clone(), v.clone()))
            .collect())
    }
}

#[async_trait]
impl<V: LiveMapValue> LiveMapFeed<String, V> for LiveMap<V> {
    async fn watch(&self, subscriber: LiveMapSubscriber<String, V>) -> Result<()> {
        let mut rx =
            self.inner.receiver.lock().unwrap().take().ok_or_else(|| {
                Error::engine("LiveMap supports a single active watch() at a time.")
            })?;
        // The framework already ran the catch-up scan + mark_ready before calling
        // `watch`. Drain incremental changes from here on. Changes that occurred
        // before the scan are replayed too; the consumer's update/delete is
        // idempotent, so the redundancy is harmless.
        while let Some(change) = rx.recv().await {
            match change {
                Change::Upsert(key, value) => subscriber.update(key, value).await?,
                Change::Delete(key) => subscriber.delete(key).await?,
            }
        }
        Ok(())
    }
}

// ---------------------------------------------------------------------------
// Producer-side target machinery: a per-map container whose children are entries.
// ---------------------------------------------------------------------------

/// Marker value for the per-map container target state.
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
struct ContainerSpec;

/// Existence marker for the container tracking record.
#[derive(Clone, Debug, Serialize, Deserialize)]
struct ContainerRecord;

#[derive(Clone, Debug, Serialize, Deserialize)]
struct ContainerAction {
    deleted: bool,
}

/// Existence marker for an entry tracking record (never read — its only job is to
/// be present so the engine can drive a delete when the producer stops declaring
/// the entry; change detection is done by `==` in the sink).
#[derive(Clone, Debug, Serialize, Deserialize)]
struct EntryRecord;

#[derive(Clone, Serialize, Deserialize)]
struct EntryAction<V> {
    key: String,
    /// `Some` to upsert, `None` to remove.
    value: Option<V>,
}

struct ContainerHandler<V> {
    inner: Arc<LiveMapInner<V>>,
}

impl<V: LiveMapValue> TargetHandler<ContainerSpec> for ContainerHandler<V> {
    type TrackingRecord = ContainerRecord;
    type Action = ContainerAction;

    fn reconcile(
        &self,
        _key: StableKey,
        desired: Option<ContainerSpec>,
        _prev: Vec<ContainerRecord>,
        _prev_may_be_missing: bool,
    ) -> Result<Option<TargetReconcileOutput<ContainerAction, ContainerRecord>>> {
        match desired {
            Some(_) => Ok(Some(TargetReconcileOutput {
                action: TargetAction::Update(ContainerAction { deleted: false }),
                sink: self.sink(),
                tracking_record: Some(ContainerRecord),
                child_invalidation: None,
            })),
            None => Ok(Some(TargetReconcileOutput {
                action: TargetAction::Delete(ContainerAction { deleted: true }),
                sink: self.sink(),
                tracking_record: None,
                child_invalidation: Some(TargetChildInvalidation::Destructive),
            })),
        }
    }
}

impl<V: LiveMapValue> ContainerHandler<V> {
    fn sink(&self) -> TargetActionSink<ContainerAction> {
        let inner = self.inner.clone();
        TargetActionSink::from_async_fn_with_children(
            move |actions: Vec<TargetAction<ContainerAction>>| {
                let inner = inner.clone();
                async move {
                    let mut out: Vec<Option<ChildTargetDef>> = Vec::with_capacity(actions.len());
                    for action in actions {
                        let a = match action {
                            TargetAction::Create(a)
                            | TargetAction::Update(a)
                            | TargetAction::Delete(a) => a,
                        };
                        if a.deleted {
                            out.push(None);
                        } else {
                            out.push(Some(ChildTargetDef::new::<V, _>(EntryHandler {
                                inner: inner.clone(),
                            })));
                        }
                    }
                    Ok(out)
                }
            },
        )
    }
}

struct EntryHandler<V> {
    inner: Arc<LiveMapInner<V>>,
}

impl<V: LiveMapValue> TargetHandler<V> for EntryHandler<V> {
    type TrackingRecord = EntryRecord;
    type Action = EntryAction<V>;

    fn reconcile(
        &self,
        key: StableKey,
        desired: Option<V>,
        _prev: Vec<EntryRecord>,
        _prev_may_be_missing: bool,
    ) -> Result<Option<TargetReconcileOutput<EntryAction<V>, EntryRecord>>> {
        let key = match &key {
            StableKey::Str(s) | StableKey::Symbol(s) => s.to_string(),
            other => return Err(Error::engine(format!("unsupported LiveMap key: {other:?}"))),
        };
        // Never skip: applying to the in-memory map is cheap and the `==` gate in
        // the sink decides whether to notify the consumer.
        match desired {
            Some(value) => Ok(Some(TargetReconcileOutput {
                action: TargetAction::Update(EntryAction {
                    key,
                    value: Some(value),
                }),
                sink: self.sink(),
                tracking_record: Some(EntryRecord),
                child_invalidation: None,
            })),
            None => Ok(Some(TargetReconcileOutput {
                action: TargetAction::Delete(EntryAction { key, value: None }),
                sink: self.sink(),
                tracking_record: None,
                child_invalidation: None,
            })),
        }
    }
}

impl<V: LiveMapValue> EntryHandler<V> {
    fn sink(&self) -> TargetActionSink<EntryAction<V>> {
        let inner = self.inner.clone();
        TargetActionSink::from_async_fn(move |actions: Vec<TargetAction<EntryAction<V>>>| {
            let inner = inner.clone();
            async move {
                for action in actions {
                    let a = match action {
                        TargetAction::Create(a)
                        | TargetAction::Update(a)
                        | TargetAction::Delete(a) => a,
                    };
                    inner.apply(a.key, a.value);
                }
                Ok(())
            }
        })
    }
}

File: python/synor/resources/live_map.py
"""In-memory live key-value map.

``LiveMap[K, V]`` bridges live data-*producing* logic and live data-*consuming* logic within a
single Synor session. The producing side declares ``(key, value)`` entries as **target
states** via :meth:`LiveMap.declare_entry`; the consuming side reads it as a
``syn.LiveMapView`` via ``syn.spawn_each``. All data is held in an in-process ``dict`` that
the engine keeps in sync through normal target-state ownership, and that same ``dict`` is
exposed as a live source for downstream components.

Designed for live mode (``app.update(live=True)``). See ``specs/live_map`` for the full design.

Example::

    lm: LiveMap[str, str] = await LiveMap.create()   # inside the app's component tree
    lm.declare_entry(key, value)                      # producer: inside any component
    await syn.spawn_each(process_entry, lm)          # consumer: one component per entry
"""

from __future__ import annotations

import asyncio as _asyncio
import uuid as _uuid
import weakref as _weakref
from collections.abc import AsyncIterator as _AsyncIterator
from collections.abc import Collection as _Collection
from collections.abc import Sequence as _Sequence
from dataclasses import dataclass as _dataclass
from typing import Any as _Any
from typing import Generic as _Generic
from typing import NamedTuple as _NamedTuple
from typing import TypeVar as _TypeVar

import msgspec as _msgspec

import synor as _syn

__all__ = ["LiveMap"]

# K becomes both the entry target-state key and the consumer's per-entry component subpath,
# so it must be a StableKey. V is any `==`-comparable value (no hashability required).
_K = _TypeVar("_K", bound="_syn.StableKey")
_V = _TypeVar("_V")

# Sentinel distinguishing "key absent" from "present with value None" in the == gate.
_MISSING: _Any = object()

# Maps a LiveMap's UUID to its live instance so the shared, stateless container sink can
# recover the Python object during apply. Weak so instances aren't retained past their
# container's teardown; a bound `_EntryHandler` holds a strong ref, keeping an in-use map
# alive (see specs/live_map/design.md).
_REGISTRY: "_weakref.WeakValueDictionary[_uuid.UUID, LiveMap[_Any, _Any]]" = (
    _weakref.WeakValueDictionary()
)


# ============================================================================
# Private target-state types
# ============================================================================


@_dataclass(frozen=True, slots=True)
class _ContainerSpec:
    """Marker value for the per-map container target state (keyed by the map's UUID)."""


class _ContainerRecord(_msgspec.Struct, frozen=True):
    """Minimal existence marker persisted for the container target state."""


class _EntryRecord(_msgspec.Struct, frozen=True):
    """Minimal existence marker persisted for an entry target state.

    Never read — its only job is to be a non-``ABSENT`` value so the engine records
    "this key exists", which lets a later run drive a delete when the producer stops declaring
    it. Change detection for the consumer is done by ``==`` in the sink, not from this record.
    """


class _ContainerAction(_NamedTuple):
    live_map: "LiveMap[_Any, _Any] | None"
    deleted: bool


class _EntryAction(_NamedTuple):
    live_map: "LiveMap[_Any, _Any]"
    key: _syn.StableKey
    value: _Any
    deleted: bool


class _Change(_NamedTuple):
    key: _Any
    value: _Any
    deleted: bool
    # Orders emissions relative to `_scan` snapshots: the watch loop skips changes
    # already reflected in the snapshot it delivered (see `watch`).
    seq: int


# ============================================================================
# Handlers and sinks (producer side)
# ============================================================================


class _EntryHandler:
    """Per-instance target handler for entries; closes over one ``LiveMap``.

    Constructed by the container sink (bound to the instance recovered from the registry) and
    reused for every entry declared into that map.
    """

    __slots__ = ("_live_map",)

    def __init__(self, live_map: "LiveMap[_Any, _Any]") -> None:
        self._live_map = live_map

    def reconcile(
        self,
        key: _syn.StableKey,
        desired_state: _Any | _syn.AbsentType,
        prev_possible_records: _Collection[_EntryRecord],
        prev_may_be_missing: bool,
        /,
    ) -> "_syn.TargetReconcileOutput[_EntryAction, _EntryRecord] | None":
        # Never skip: applying to an in-memory dict is cheap, and the `==` gate in the sink
        # decides whether to notify the consumer. `prev_possible_records` is intentionally
        # unused (no fingerprint compare).
        if _syn.is_absent(desired_state):
            return _syn.TargetReconcileOutput(
                action=_EntryAction(self._live_map, key, None, True),
                sink=_ENTRY_SINK,
                tracking_record=_syn.ABSENT,
            )
        return _syn.TargetReconcileOutput(
            action=_EntryAction(self._live_map, key, desired_state, False),
            sink=_ENTRY_SINK,
            tracking_record=_EntryRecord(),
        )


class _ContainerHandler:
    """Root handler: binds a per-instance ``_EntryHandler`` via the UUID→instance registry."""

    def reconcile(
        self,
        key: _syn.StableKey,
        desired_state: "_ContainerSpec | _syn.AbsentType",
        prev_possible_records: _Collection[_ContainerRecord],
        prev_may_be_missing: bool,
        /,
    ) -> "_syn.TargetReconcileOutput[_ContainerAction, _ContainerRecord, _EntryHandler] | None":
        if _syn.is_absent(desired_state):
            return _syn.TargetReconcileOutput(
                action=_ContainerAction(None, True),
                sink=_CONTAINER_SINK,
                tracking_record=_syn.ABSENT,
            )
        assert isinstance(key, _uuid.UUID)
        return _syn.TargetReconcileOutput(
            action=_ContainerAction(_REGISTRY.get(key), False),
            sink=_CONTAINER_SINK,
            tracking_record=_ContainerRecord(),
        )


async def _apply_entry_actions(
    context_provider: _syn.ContextProvider,
    actions: _Sequence[_EntryAction],
    /,
) -> None:
    # Runs on the app event loop (async sink), so it mutates `_entries` and notifies the
    # watcher queue directly — no locks, no cross-thread marshalling.
    for action in actions:
        live_map = action.live_map
        if action.deleted:
            if action.key in live_map._entries:
                del live_map._entries[action.key]
                live_map._seq += 1
                live_map._emit(_Change(action.key, None, True, live_map._seq))
        else:
            prev = live_map._entries.get(action.key, _MISSING)
            if prev is _MISSING or prev != action.value:
                live_map._entries[action.key] = action.value
                live_map._seq += 1
                live_map._emit(_Change(action.key, action.value, False, live_map._seq))


async def _apply_container_actions(
    context_provider: _syn.ContextProvider,
    actions: _Sequence[_ContainerAction],
    /,
) -> "list[_syn.ChildTargetDef[_EntryHandler] | None]":
    out: "list[_syn.ChildTargetDef[_EntryHandler] | None]" = []
    for action in actions:
        if action.deleted or action.live_map is None:
            out.append(None)
        else:
            out.append(_syn.ChildTargetDef(handler=_EntryHandler(action.live_map)))
    return out


_ENTRY_SINK: _syn.TargetActionSink[_EntryAction, None] = (
    _syn.TargetActionSink.from_async_fn(_apply_entry_actions)
)
_CONTAINER_SINK: _syn.TargetActionSink[_ContainerAction, _EntryHandler] = (
    _syn.TargetActionSink.from_async_fn(_apply_container_actions)
)
_CONTAINER_PROVIDER = _syn.register_root_target_states_provider(
    "synor/livemap", _ContainerHandler()
)


# ============================================================================
# Public API
# ============================================================================


class LiveMap(_Generic[_K, _V]):
    """An in-memory, keyed collection that is both a target and a ``syn.LiveMapView``.

    Create with ``await LiveMap.create()`` from inside the app's component tree. Producers add
    entries with :meth:`declare_entry` from inside any component; consumers pass the map to
    ``syn.spawn_each`` to process one component per entry, kept in sync as entries appear,
    change, and disappear. An entry exists as long as some live component declares it; when its
    declaring component stops declaring it (or disappears), the entry is removed.

    Single active ``watch()`` at a time. ``K`` must be a ``syn.StableKey``; ``V`` is any
    ``==``-comparable value (no hashability required).
    """

    __slots__ = (
        "_uuid",
        "_entries",
        "_entry_provider",
        "_watcher_queue",
        "__weakref__",
        "_seq",
        "_watch_scan_seq",
    )

    _uuid: _uuid.UUID
    _entries: dict[_K, _V]
    _entry_provider: "_syn.TargetStateProvider[_Any, None]"
    _watcher_queue: "_asyncio.Queue[_Change] | None"

    def __init__(self) -> None:
        # Only sets up in-memory state — `create()` mounts the container and sets
        # `_entry_provider`; the map isn't usable until then.
        self._uuid = _uuid.uuid4()
        self._entries = {}
        self._watcher_queue = None
        # Monotonic emission counter (single event loop: applies are sequential, so a
        # change with seq <= the seq recorded at snapshot time is already in that
        # snapshot).
        self._seq = 0
        self._watch_scan_seq: int | None = None

    @classmethod
    async def create(cls) -> "LiveMap[_K, _V]":
        """Create a LiveMap and mount its backing target. Call inside a component context."""
        self = cls()
        _REGISTRY[self._uuid] = self
        container = _CONTAINER_PROVIDER.target_state(self._uuid, _ContainerSpec())
        self._entry_provider = await _syn.attach_target(container)
        return self

    def declare_entry(self, key: _K, value: _V) -> None:
        """Declare an entry, owned by the calling component. Call inside a component context."""
        _syn.ensure_target_state(self._entry_provider.target_state(key, value))

    def __aiter__(self) -> _AsyncIterator[tuple[_K, _V]]:
        return self._scan()

    async def _scan(self) -> _AsyncIterator[tuple[_K, _V]]:
        # Snapshot synchronously so a sink firing between yields can't mutate mid-iteration.
        snapshot = list(self._entries.items())
        if self._watcher_queue is not None and self._watch_scan_seq is None:
            # First scan after a watch armed its queue = the watcher's initial
            # snapshot (`subscriber.update_all`): record how far it reached.
            self._watch_scan_seq = self._seq
        for item in snapshot:
            yield item

    async def watch(self, subscriber: "_syn.LiveMapSubscriber[_K, _V]") -> None:
        """Deliver an initial snapshot then incremental changes. Drives one consumer."""
        if self._watcher_queue is not None:
            raise RuntimeError("LiveMap supports a single active watch() at a time.")
        queue: _asyncio.Queue[_Change] = _asyncio.Queue()
        # Arm before the scan so changes concurrent with it aren't lost. The mirror
        # image of that choice is a change landing between arming and the snapshot:
        # it gets queued AND included in the snapshot. The seq gate below drops such
        # already-delivered changes at drain time (they'd otherwise re-notify the
        # consumer with an equal value, defeating the `==` gate).
        self._watcher_queue = queue
        self._watch_scan_seq = None
        try:
            await subscriber.update_all()
            await subscriber.mark_ready()
            snapshot_seq = self._watch_scan_seq
            while True:
                change = await queue.get()
                if snapshot_seq is not None and change.seq <= snapshot_seq:
                    continue  # already reflected in the initial snapshot
                if change.deleted:
                    handle = await subscriber.delete(change.key)
                else:
                    handle = await subscriber.update(change.key, change.value)
                await handle.ready()
        finally:
            self._watcher_queue = None
            self._watch_scan_seq = None

    def _emit(self, change: _Change) -> None:
        queue = self._watcher_queue
        if queue is not None:
            queue.put_nowait(change)
